plugin/src/main/java/io/snyk/eclipse/plugin/views/snyktoolview/TreeViewerFilter.java

```diff
@@ -1,55 +1,66 @@
 package io.snyk.eclipse.plugin.views.snyktoolview;
 
-import java.util.ArrayList;
-import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
 
-import org.eclipse.jface.viewers.ILabelProvider;
+import org.eclipse.jface.viewers.ITreeContentProvider;
 import org.eclipse.jface.viewers.TreeViewer;
 import org.eclipse.jface.viewers.Viewer;
 import org.eclipse.jface.viewers.ViewerFilter;
 
+import io.snyk.languageserver.protocolextension.messageObjects.scanResults.Issue;
+
 public class TreeViewerFilter extends ViewerFilter {
-	private List<String> searchStrings;
-	private boolean matchAll; // If true, all filters must match; if false, any filter can match
+	private Map<String, Predicate<? super Issue>> filters;
 
 	public TreeViewerFilter() {
-		this.searchStrings = new ArrayList<>();
-		this.matchAll = true; // Default to matching all filters
+		filters = new ConcurrentHashMap<>();
 	}
 
-	public void addSearchText(String s) {
-		if (s != null && !s.isEmpty()) {
-			this.searchStrings.add(".*" + s.toLowerCase() + ".*");
-		}
-	}
+	@Override
+	public boolean select(Viewer viewer, Object parentElement, Object element) {
+	    if (element instanceof FileTreeNode) {
+	        return hasVisibleChildren((FileTreeNode) element, viewer);
```

Comment:
that looks good. I hope it wouldn't be a performance bottleneck tho.

Reply:
Agreed, we only call the ```hasVisibleChildren``` if the node is a treenode, so its not called for all tree nodes. However this might be slow if a project has very many files.